### PR TITLE
Switch FmtSpan To a Combinable Bitflag

### LINKED ISSUE #1136: Allow for synthesized events on created/closed
### Crates

* tracing_subscriber

### Motivation

The typical example for a span is something like an HTTP request. If a span is performed in this way, it may be desirable to have the span synthesize an event. I could see using `FmtSpan::CLOSE`, which gives you a single event for the entire request. However, if I want two requests (one at the start, one at the end) there aren't really good options: both `ACTIVE` and `FULL` will give me an event at the start of the request, but potentially even more if I am polling some futures during the request handling.

### Proposal

Add another item to the enum, `FmtSpan::???`. When used, synthesized events will be emitted on only new and drop events.

### Alternatives

Or perhaps I'm not understanding the full purpose of the `FmtSpan` type, and it's maybe better in this case to use some other way of filtering this out?

### PR DESCRIPTION
**Pull request by zicklag**
Fixes #1136.

Allows arbitrarily combining different FmtSpan events to listen to.

## Motivation

The idea is to allow any combination of `FmtSpan` options, such as the currently unachievable combination of `FmtSpan::NEW | FmtSpan::CLOSE`.

## Solution

Make `FmtSpan` behave like a bitflag that can be combined using the bitwise or operator ( `|` ) while maintaining backward compatibility by keeping the same names for all existing presets and keeping the implementation details hidden.

### CODE REVIEW

**Comment by hawkw** [comment]
Great, thanks for working on this! 

I'm going to give a more complete review shortly, but before I do, I just wanted to let you know that the CI failure is due to [an issue with a dependency](https://github.com/tokio-rs/tracing/pull/1275#issuecomment-789592474) and is not your fault or related to this change at all. Fortunately it looks like that'll be fixed upstream soon.

File: tracing-subscriber/src/fmt/fmt_subscriber.rs
Comment (davidbarsky) [inline tracing-subscriber/src/fmt/fmt_subscriber.rs:248]: I _personally_ dislike using the second person in documentation ("you"). How about:

```suggestion
    /// The options can be enabled in any combination. For instance, the following
    /// will synthesize events whenever spans are created and closed:
```

File: tracing-subscriber/src/fmt/fmt_subscriber.rs
Comment (davidbarsky) [inline tracing-subscriber/src/fmt/fmt_subscriber.rs:234]:
```suggestion
    /// - `FmtSpan::NEW`: An event will be synthesized when spans are created.
    /// - `FmtSpan::ENTER`: An event will be synthesized when spans are entered.
    /// - `FmtSpan::EXIT`: An event will be synthesized when spans are exited.
```

File: tracing-subscriber/src/fmt/mod.rs
Comment (davidbarsky) [inline tracing-subscriber/src/fmt/mod.rs:647]:
```suggestion
    /// - `FmtSpan::NEW`: An event will be synthesized when spans are created.
    /// - `FmtSpan::ENTER`: An event will be synthesized when spans are entered.
    /// - `FmtSpan::EXIT`: An event will be synthesized when spans are exited.
```

File: tracing-subscriber/src/fmt/mod.rs
Comment (davidbarsky) [inline tracing-subscriber/src/fmt/mod.rs:654]:
```suggestion
    /// - `FmtSpan::ACTIVE`: An event will be synthesized when spans are entered
    ///    or exited.
```

Also, question for @hawkw: should we deprecate this variant and pushing folks to use `FmtSpan::Enter | FmtSpan::Exit` instead?

File: tracing-subscriber/src/fmt/mod.rs
Comment (davidbarsky) [inline tracing-subscriber/src/fmt/mod.rs:661]: Same thing about the second person usage. Sorry!

```suggestion
    /// The options can be enabled in any combination. For instance, the following
    /// will synthesize events whenever spans are created and closed:
```

**Review by davidbarsky — CHANGES_REQUESTED** [review]

**Comment by davidbarsky** [comment]
Thanks for opening this PR! Just a few comments.

**Comment by zicklag** [comment]
No problem, glad I could help! 

I committed all of your wording suggestions.

File: tracing-subscriber/src/fmt/mod.rs
Comment (davidbarsky) [reply tracing-subscriber/src/fmt/mod.rs]: (These changes have been merged, but unresolving to make this discussion more visible.)

Comment (zicklag) [reply tracing-subscriber/src/fmt/mod.rs]: IMHO `ACTIVE` is not the most intuitive name and it would be more natural if you just combined them explicitly. That would make things more "self-documenting" with the code being more obvious and longer descriptions of the meaning of flags less necessary.

File: tracing-subscriber/src/fmt/format/mod.rs
Comment (hawkw) [inline tracing-subscriber/src/fmt/format/mod.rs:1113]: If multiple bitflags are set, this looks like it will just concatenate them in a kind of ugly way. for example, suppose NEW, ENTER, and EXIT flags are set, this will format that as `FmtSpan::NEWFmtSpan::ENTERFmtSpan::EXIT`. I think we should maybe delimit them with `|` instead, maybe like this:
```suggestion
        let mut wrote_flags = false;
        if self.contains(FmtSpan::NEW) {
            f.write_str("FmtSpan::NEW")?;
            wrote_flags = true;
        }
        
        if self.contains(FmtSpan::ENTER) {
            if wrote_flags {
                f.write_str(" | ")?;
            }
            f.write_str("FmtSpan::ENTER")?;
            wrote_flags = true;
        }
        
        if self.contains(FmtSpan::EXIT) {
            if wrote_flags {
                f.write_str(" | ")?;
            }
            f.write_str("FmtSpan::EXIT")?;
            wrote_flags = true;
        }
        
        if self.contains(FmtSpan::CLOSE) {
            if wrote_flags {
                f.write_str(" | ")?;
            }
            f.write_str("FmtSpan::CLOSE")?;
            wrote_flags = true;
        }
```

we could probably also write a helper function to make that a little less repetitive, if you like...

Comment (hawkw) [reply tracing-subscriber/src/fmt/format/mod.rs:1113]: we might also want to add
```rust
if !wrote_flags {
    f.write_str("FmtSpan::NONE")?;
}
```
at the very end, so that an empty set of flags doesn't format itself as the empty string?

File: tracing-subscriber/src/fmt/format/mod.rs
Comment (hawkw) [inline tracing-subscriber/src/fmt/format/mod.rs:1094]: i'm not totally convinced we actually *need* an `&` operation, as i'm not sure how often user code will actually need to test for the presence of flags. i don't think providing it is a *problem*, though, so this is fine

**Review by hawkw — CHANGES_REQUESTED** [review]
Overall this looks good to me!

I noticed that the `fmt::Debug` implementation for `FmtSpan` will produce somewhat ugly output, so I suggested a potential improvement there. Once that's fixed, I'll be happy to merge this!

File: tracing-subscriber/src/fmt/format/mod.rs
Comment (zicklag) [reply tracing-subscriber/src/fmt/format/mod.rs:1094]: Yeah, I wasn't positive either, but I figured if we were going to act like it was a bitflag, a user might _expect_ that they could do it so I thought we might as well add it.

File: tracing-subscriber/src/fmt/format/mod.rs
Comment (zicklag) [reply tracing-subscriber/src/fmt/format/mod.rs:1113]: Whoops, that's a good point, I'll fix it.

File: tracing-subscriber/src/fmt/format/mod.rs
Comment (hawkw) [inline tracing-subscriber/src/fmt/format/mod.rs:1123]: <3 looks great

**Review by hawkw — APPROVED** [review]

File: tracing-subscriber/src/fmt/format/mod.rs
Comment (hawkw) [inline tracing-subscriber/src/fmt/format/mod.rs:1341]: TIOLI: it could be nice to have a test that we actually do reasonable behavior with various bitflag combinations, as well, but I'm not going to block merging this on that, the code is pretty straightforward.

**Review by hawkw — APPROVED** [review]
I'm going to go ahead and merge this now; we can add more tests etc in a follow-up.

Thanks again!

**Comment by zicklag** [comment]
Glad I could help! This was one of my pet annoyances with tracing so it's cool to be able to fix it. :smile:

**Pull request merged**
